GVFS/GVFS.Common/NuGetUpgrade/ModernNuGetUpgrader.cs

```diff
@@ -0,0 +1,256 @@
+            }
+            catch (TaskCanceledException exception)
+            {
+                // GetStringAsync can also throw a TaskCanceledException to indicate a timeout
+                // https://github.com/dotnet/corefx/issues/20296
+                message = string.Format("Network error: could not connect to server ({0}). {1}", this.OrgInfoServerUrl, exception.Message);
+                this.TraceException(exception, nameof(this.TryQueryNewestVersion), $"Error connecting to server.");
+
+                return false;
+            }
+            catch (SerializationException exception)
+            {
+                message = string.Format("Parse error: could not parse response from server({0}). {1}", this.OrgInfoServerUrl, exception.Message);
+                this.TraceException(exception, nameof(this.TryQueryNewestVersion), $"Error parsing response from server.");
+
+                return false;
+            }
+
+            if (this.highestVersionAvailable != null &&
+                this.highestVersionAvailable > this.installedVersion)
+            {
+                newVersion = this.highestVersionAvailable;
+            }
+
+            if (newVersion != null)
+            {
+                this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - new version available: installedVersion: {this.installedVersion}, highestVersionAvailable: {newVersion}");
+                message = $"New version {this.highestVersionAvailable} is available.";
+                return true;
+            }
+            else if (this.highestVersionAvailable != null)
+            {
+                this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - up-to-date");
+                message = $"highest version available is {this.highestVersionAvailable}, you are up-to-date";
+                return true;
+            }
+            else
+            {
+                this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - no versions available from feed.");
+                message = $"No versions available via endpoint.";
+            }
+
+            newVersion = this.highestVersionAvailable;
+            message = null;
+            return true;
+        }
+
+        public class ModernNuGetUpgraderConfig : NuGetUpgraderConfig
+        {
+            public ModernNuGetUpgraderConfig(ITracer tracer, LocalGVFSConfig localGVFSConfig)
+                : base(tracer, localGVFSConfig)
+            {
+            }
+
+            public string OrgInfoServer { get; set; }
+
+            public string UpgradeRing { get; set; }
+
+            public override bool TryLoad(out string error)
+            {
+                if (!base.TryLoad(out error))
+                {
+                    return false;
+                }
+
+                if (!this.localConfig.TryGetConfig(GVFSConstants.LocalGVFSConfig.OrgInfoServerUrl, out string orgInfoServerUrl, out error))
+                {
+                    return false;

```

Comment:
Do we need to log anything here or when returning `false`?  Looks suspicious only logging in the one case.